man/man1/ckmf.1
.TH CKMF 1
.\" @(#) $Header: /xtel/pp/pp-beta/man/man1/RCS/ckmf.1,v 6.0 1991/12/18 20:43:41 jpo Rel $
.\"
.\" $Log: ckmf.1,v $
.\" Revision 6.0  1991/12/18  20:43:41  jpo
.\" Release 6.0
.\"
.\"
.\"
.SH NAME
ckmf \- check users .mailfilter file
.SH SYNOPSIS
.in +.5i
.ti -.5i
ckmf file
.in -.5i
.SH DESCRIPTION
\fICkmf\fP checks that the given .mailfilter is free from syntax
errors. This does not necessarily mean that the file will work, just
that it should be interpreted ok. Run time errors may still occur.
.PP
It runs with a single argument, the file to check. It expects a
message on the standard input to attempt to deliver. It will then
print out what actions it would take. An empty message (a.k.a.
/dev/null) may be given and the appropriate actons will be indicated.
.PP
It is \fBSTRONGLY\fP advised that you run this program after modifying
your \&\fB.mailfilter\fP file to check that you haven't made mistakes
in the syntax.
.SH FILES
.nf
.ta \w'/usr/local/lib/mailfilter\0\0'u
~/.mailfilter	The users mailfilter file
/usr/local/lib/mailfilter	The system mailfilter file
.fi
.SH "SEE ALSO"
\fIPP Manual \- Volume 3: User Manual\fP
.SH BUGS
Ought to check more (including the sanity of the user).


CHANGES









             Changes from PP version 5.2 to 6.0





_1.  _O_v_e_r_v_i_e_w

This  short  document  is  a   description   of   the   more
user/administrator  visible changes that have taken place in
PP between the last interim version (5.2) and the beta  ver-
sion  (6.0). This document is not exhaustive or complete but
is a guide to what to look out for. In general the following
may be assumed.

o+    The Queue files are compatible  between  5.2  and  6.0.
     Note  however that the Queue structure has changed. Old
     channels will have to be recompiled  to  recognise  the
     new format, but new channels will read old queue files.

o+    The qmgr protocol is compatible between 5.2 and 6.0

o+    The syntax of some of  the  table  files  has  changed.
     Where appropriate conversion scripts are provided.

o+    Some procedure calls have been  changed  or  have  dif-
     ferent parameters.

o+    the libdsap.a library is now required for compilation.

In summary, you should be careful when upgrading to the  6.0
release.   It  will  not  work  by  just  replacing  the 5.2
binaries with 6.0. For definite the domain table  will  HAVE
to be changed.

_2.  _D_e_t_a_i_l_e_d _c_h_a_n_g_e_s

This list is in no particular order of preference or  impor-
tance.

Domain Table
     The domain table syntax has been  changed  to  be  more
     explicit  and  has  been  brought into line with the OR
     table syntax. We feel this is clearer and more logical.
     Tools  that  process  the domain table have been taught
     the   new   format   and   there   are    scripts    in
     Tools/tables/compat  to convert between the old and new
     formats.  There is NO backwards compatibility here. Old
     format tables will just plain not work.

Aliases Table



                     December 18, 1991





                           - 2 -


     The aliases table syntax has changed to a  main  choice
     of  synonym or alias with further qualifiers specifying
     the address type of the synonym or alias.  The  changes
     are

          x:alias y         x:alias y
          x:synonym y       x:synonym y
          x:822-synonym y   x:synonym y 822
          x:x400-synonym y  x:synonym y x400
          x:822 y           x:alias y 822
          x:x400 y          x:alias y x400
     where any white space in  y  must  be  quoted  using  a
     backslash  There  is also a "external" qualifier.  This
     disables further lookups of synonyms  and  aliases  and
     can  be  used  for  a  main  site mail hub mapping from
     site_wide_name@localsite                             to
     local_machine_name@localmachine  and  vice  versa.  The
     use of external stops the looping alias  problem  which
     would otherwise occur.

Return of Contents
     Return of contents is now supported.  This  is  a  side
     effect  and  part  of  the  work on routing of delivery
     reports.

Use of pepsy
     Use of the PEPSY ASN.1 compiler rather  than  PEPY/POSY
     is  now  used  in  many  places. This makes for smaller
     binaries and much quicker compilation.

ISODE 7.0
     PP 6.0 is aligned to ISODE 7.0. It will not work  under
     ISODE 6.0.  If you are using the 6.0 release you should
     be using ISODE 7.0. It should however work under  ISODE
     6.8.

Tailor file
     There have been many changes to the  tailor  file.  The
     introduction of several new variables and the splitting
     up of header and body part types. These are

          dap_user     Directory username for DS lookups
          dap_passwd   Directory Passwd for same
          fsync        disable use of fsync(2)
          headertype   list of headers c.f. bodypart
          queuestruct  use hierarchical queue structure
          submit_addr  call submit over network
          diskuse      limit the disk space consumed by PP

     Channel tailoring also has a number of new variables

          drchan       replaces old outchan
          hdrin        headers inbound
          hdrout       headers outbound



                     December 18, 1991





                           - 3 -


          ininfo       inbound channel info
          inadr        inbound channel addr type
          insubadr     inbound channel sub address type
          intable      inbound channel table
          key          give alternative names to channel
          mtatable     list of known mtas for channel binding
          outinfo      outbound channel info
          outsubadr    outbound channel sub address
          outtable     outbound channel table
          solo-proc    channel requires address one at a time
          bad-sender-policyset the policy for unroutable senders
          check        set the checking mode of the channel

Channel pairing
     Channels may now be explicitly  paired  together.  This
     allows one definition to specify both an inbound and an
     outbound channel.  It also allows several logical chan-
     nels  for  one  inbound protocol by selecting a channel
     per MTA.

Timezones
     Many more lettered timezones are recognised. This  list
     is ambiguous at best and numeric forms should really be
     used except at user interfaces.

Delivery Reports
     Delivery reports are now routed back to their  origina-
     tor.  In 5.2 they were returned directly down the chan-
     nel they arrived on. This allows for one way  links  in
     general  makes  things  easier. It also allows protocol
     conversion so that, for instance, a delivery report  on
     an  X.400 message may go back via an RFC-822 if that is
     the only option.  A side effect of this is the  channel
     variable  outchan  has been renamed drchan and has dif-
     ferent semantics attached to it (see the manual).

Header parts
     Separation of headers and bodyparts for  each  channel.
     This  is  mainly a syntactic change but will allow more

Chans/822-local/ckmf.c
/* chkmf.c: check local delivery file stuff */

# ifndef lint
static char Rcsid[] = "@(#)$Header: /xtel/pp/pp-beta/Chans/822-local/RCS/ckmf.c,v 6.0 1991/12/18 20:04:31 jpo Rel $";
# endif

/*
 * $Header: /xtel/pp/pp-beta/Chans/822-local/RCS/ckmf.c,v 6.0 1991/12/18 20:04:31 jpo Rel $
 *
 * $Log: ckmf.c,v $
 * Revision 6.0  1991/12/18  20:04:31  jpo
 * Release 6.0
 *
 */



#include "util.h"
#include <varargs.h>
#include "expand.h"
#include "retcode.h"
#include <sys/stat.h>
#include <pwd.h>
#include <isode/tailor.h>	/* getlocalhost defn */

#define PROTMODE	(0022)

#define MAXVARS 100
Expand	variables[MAXVARS];
int	nvars;
static int fakeuserid;
static int change_user = 0;

char	*myname;
extern int debug, yydebug;
void	adios (), advise ();

main (argc, argv)
int	argc;
char	**argv;
{
	extern char	*optarg;
	extern int	optind;
	extern FILE	*yyin;
	int	opt;
	int	n;
	char	msg_size[20];
	struct stat st;

	myname = argv[0];
	while((opt = getopt(argc, argv, "du:i:")) != EOF)
		switch (opt) {
		    case 'd':
			yydebug = debug = 1;
			break;

		    case 'u':
			{
				struct passwd *pwd;

				if ((pwd = getpwnam (optarg)) == NULL) {
					fprintf (stderr,
						 "Unknown user name %s",
						 optarg);
					exit (1);
				}
				fakeuserid = pwd -> pw_uid;
				change_user = 1;
			}
			break;
		    case 'i':
			fakeuserid = atoi(optarg);
			change_user = 1;
			break;

		    default:
			fprintf (stderr, "Usage: %s [-d] mailfilter\n",
				 myname);
			break;
		}
	argc -= optind;
	argv += optind;
	if (change_user)
		do_the_change (fakeuserid);

	if (argc != 1)
		adios (NULLCP, "Usage: %s [-d] mailfilter", myname);
	if ((yyin = fopen (argv[0], "r")) == NULL)
		adios (argv[0], "Can't open file");
	if (fstat (fileno (yyin), &st) == NOTOK)
		adios (argv[0], "Can't fstat file");
	if (st.st_mode & PROTMODE) {
		fprintf (stderr, "File %s has wrong mode (%o)\n",
			 argv[0], st.st_mode & 0777);
		fprintf (stderr,
			 "Should not have group or general write permission\n");
	}
		
	initialise ();

	yyparse();

	if (fstat (0, &st) == -1)
		(void) strcpy (msg_size, "???");
	else	(void) sprintf (msg_size, "%d", st.st_size);

	n = 0;
	variables[n].macro = strdup("size");
	variables[n].expansion = strdup (msg_size);
	create_var (&variables[n]);
	n++;

	variables[n].macro = strdup ("return-path");
	variables[n].expansion = strdup ("sender@remote");
	create_var (&variables[n]);
	n++;

	variables[n].macro = strdup ("mailbox");
	variables[n].expansion = strdup ("themailbox");
	create_var (&variables[n]);
	n++;

	variables[n].macro = strdup ("recipient");
	variables[n].expansion = strdup ("therecipient");
	create_var (&variables[n]);
	n++;

	variables[n].macro = strdup ("userid");
	if (getenv ("USER"))
		variables[n].expansion = strdup (getenv("USER"));
	else {
		(void) sprintf (msg_size, "%d", getuid ());
		variables[n].expansion = strdup (msg_size);
	}
	create_var (&variables[n]);
	n ++;

	(void) sprintf (msg_size, "%d", getgid ());
	variables[n].macro = strdup ("groupid");
	variables[n].expansion = strdup (msg_size);
	create_var (&variables[n]);
	n ++;

	variables[n].macro = strdup ("channelname");
	variables[n].expansion = strdup ("thechannel");
	create_var (&variables[n]);
	n++;

	variables[n].macro = strdup ("hostname");
	variables[n].expansion = strdup (getlocalhost ());
	create_var (&variables[n]);
	n ++;
	

	nvars = parse_hdr (0, variables + 3, MAXVARS - 3) + 3;

	switch (run ()) {
	    case RP_OK:
		printf ("Result: delivered OK\n");
		break;
	    case RP_BAD:
		printf ("Result: delivery failed (permanently)\n");
		break;
	    case RP_AGN:
		printf ("Result: delivered failed (temporarily)\n");
		break;
	    case RP_NOOP:
		printf ("Result: no delivery - other methods will be tried\n");
		break;
	    default:
		printf ("Result: Dunno!\n");
		break;
	}

	exit (0);
	/* NOTREACHED */
}

do_the_change (id)
int	id;
{
	int myuid = getuid ();
	struct passwd *pwd;
	extern char *pplogin;

	pwd = getpwnam (pplogin);
	if (myuid == 0 ||
	    (pwd && myuid == pwd -> pw_uid)) {
#ifdef BSD42
		if (setreuid (id, id) == NOTOK)
#else
		if (setuid (id) == NOTOK)
#endif
			adios ("setreuid", "Can't set ids to %d", id);
	}
	else
		adios (NULLCP, "Not running as root or PP");
}


#ifndef	lint
static void	_advise ();


void	adios (va_alist)
va_dcl
{
    va_list ap;

    va_start (ap);

    _advise (ap);

    va_end (ap);

    _exit (1);
}
#else
/* VARARGS */

void	adios (what, fmt)
char   *what,
       *fmt;
{
    adios (what, fmt);
}
#endif


#ifndef	lint
void	advise (va_alist)
va_dcl
{
    va_list ap;

    va_start (ap);

    _advise (ap);

    va_end (ap);
}


static void  _advise (ap)
va_list	ap;
{
    char    buffer[BUFSIZ];

    asprintf (buffer, ap);

    (void) fflush (stdout);

    fprintf (stderr, "%s: ", myname);
    (void) fputs (buffer, stderr);
    (void) fputc ('\n', stderr);

    (void) fflush (stderr);
}
#else
/* VARARGS */

void	advise (what, fmt)
char   *what,
       *fmt;
{
    advise (what, fmt);
}
#endif

putunixfile (str)
char	*str;
{
	char	buf[BUFSIZ];
	expand (buf, str, variables);
	printf ("append to file (sendmail format) '%s'\n", buf);
	return RP_OK;
}

putfile (str)
char	*str;
{
	char	buf[BUFSIZ];
	expand (buf, str, variables);
	printf ("append to file '%s'\n", buf);
	return RP_OK;
}

putpipe (str)
char	*str;
{
	char	buf[BUFSIZ];

	install_vars (variables, nvars, MAXVARS);
	expand (buf, str, variables);
	printf ("pipe into process '%s'\n", buf);
	return RP_OK;
}

printit (s)
char	*s;
{
	fputs (s, stdout);
}
